torch_geometric/nn/conv/gatv2_conv.py

```diff
@@ -147,13 +193,31 @@ def forward(self, x: Union[Tensor, PairTensor], edge_index: Adj,
                     num_nodes = min(num_nodes, x_r.size(0))
                 if size is not None:
                     num_nodes = min(size[0], size[1])
-                edge_index, _ = remove_self_loops(edge_index)
-                edge_index, _ = add_self_loops(edge_index, num_nodes=num_nodes)
+                edge_index, edge_attr = remove_self_loops(
+                    edge_index, edge_attr)
+                edge_index, edge_attr = add_self_loops(
+                    edge_index, edge_attr, fill_value=self.fill_value,
+                    num_nodes=num_nodes)
             elif isinstance(edge_index, SparseTensor):
-                edge_index = set_diag(edge_index)
+                if self.edge_dim is None:
+                    edge_index = set_diag(edge_index)
+                else:
+                    raise NotImplementedError(
+                        "The usage of 'edge_attr' and 'add_self_loops' "
+                        "simultaneously is currently not yet supported for "
+                        "'edge_index' in a 'SparseTensor' form")
+
+        edge_attr_transformed: OptTensor = None
+        if edge_attr is not None:
+            if edge_attr.dim() == 1:
+                edge_attr = edge_attr.view(-1, 1)
+            assert self.lin_edge is not None
+            edge_attr_transformed = self.lin_edge(edge_attr)
+            edge_attr_transformed = edge_attr_transformed.view(-1, H, C)
```

Comment:
This is a minor nitpick. Do you think we can move this piece of code into the `message` function, this makes it more consistent with other layers that use `edge_attr` like [TransformerConv](https://github.com/pyg-team/pytorch_geometric/blob/958b1dd3e80e0ea670f1cb149e536ed1e98b55cc/torch_geometric/nn/conv/transformer_conv.py#L220) or [GatConv](https://github.com/pyg-team/pytorch_geometric/blob/958b1dd3e80e0ea670f1cb149e536ed1e98b55cc/torch_geometric/nn/conv/gat_conv.py#L256).

Reply:
Thanks for the feedback!
I moved it to the message function for consistency :)
However, I found it a little easier to understand in the forward function, since the node features (x) are also usually transformed there.
Do you know why this is handled differently for the edge features? Does it have performance benefits?

Reply:
Given the message->aggregate->update scheme, ideally both the linear operation on node and linear on edge attributes should be computed in the `message` function since these operations are used to compute the message.

But running the linear operation in forward for the node features gives you performance benefits, as in message the operation is over x_i or x_j(E*F) and in forward its just over x(N*F). For edge attributes it where the linear operation is done doesn't make a difference.

Reply:
I was a bit confused because forward and message are both only called once in the current setup and I thought there would be no performance differences. But now I realise that MessagePassing's propagate is encapsulating the rearrangement of the elements of X that are passed to the message function.
Thank you :)